tests/handlers/test_sync.py

```diff
@@ -186,6 +187,139 @@ def test_unknown_room_version(self):
+        * racing with that, is a ban which comes in over federation, which ends up with
+          an earlier stream_ordering than the join.
+        * you get a sync response with a sync token which is _after_ the ban
+        * now your join lands; it is a valid event because its `prev_event`s predate the
+          ban, but will not make it into current_state_events (because bans win over
+          joins in state res, essentially).
+        * Hence, the only event in the timeline is your join ... and yet you aren't
+          joined.
+        """
+        # A local user Alice creates a room.
+        owner = self.register_user("alice", "password")
+        owner_tok = self.login(owner, "password")
+        room_id = self.helper.create_room_as(owner, is_public=True, tok=owner_tok)
+
+        # Do a sync as Alice to get the latest event in the room.
+        alice_sync_result: SyncResult = self.get_success(
+            self.sync_handler.wait_for_sync_for_user(
+                create_requester(owner), generate_sync_config(owner)
+            )
+        )
+        self.assertEqual(len(alice_sync_result.joined), 1)
+        self.assertEqual(alice_sync_result.joined[0].room_id, room_id)
+        last_room_creation_event_id = (
+            alice_sync_result.joined[0].timeline.events[-1].event_id
+        )
+
+        # (Pretend that) a remote user Bob joins.
+        moderator = "@bob:example.com"
+        self.get_success(
+            inject_event(
+                self.hs,
+                sender=moderator,
+                prev_event_ids=[last_room_creation_event_id],
+                type=EventTypes.Member,
+                state_key=moderator,
+                room_id=room_id,
+                content={"membership": "join"},
+            )
+        )
+
+        # Alice makes Bob a moderator.
+        room_power_levels = self.helper.get_state(
+            room_id,
+            EventTypes.PowerLevels,
+            tok=owner_tok,
+        )
+        room_power_levels["users"][moderator] = 50
+        moderator_pl_event_id = self.helper.send_state(
+            room_id,
+            EventTypes.PowerLevels,
+            room_power_levels,
+            tok=owner_tok,
+        )["event_id"]
+
+        # We now want to concoct a situation where:
+        # - another local user Eve attempts to join the room
+        # - the homeserver generates a join event with prev_events that precede the ban
+        #   (so that it passes the "are you banned" test)
+        # - but the join event has a stream_ordering after that of the ban (so that it
+        #   comes down /sync after the ban).
+        #
+        # We do this by
+        # 1. Ban eve from the room.
+        # 2. Have eve join the room.
+        #    - mock the prev_events for the join event to precede the ban
+
+        # First, the ban.
+        eve = self.register_user("eve", "password")
+        eve_token = self.login(eve, "password")
+        self.get_success(
```

Comment:
Can this just be a normal `send_state`?

Reply:
This one is coming from the moderator, who we're pretending is on a remote homeserver. But on reflection, I don't think it matters who sends the ban or where their homeserver is. (The scenario Rich describes involving federation is just one way that it could happen).

Maybe there's no need for the `moderator` user at all---maybe the `owner` can do the banning. I'll have a look at this tomorrow!

Reply:
Yeah, the federation example might just be a likely way for this to happen, but perhaps something similar could happen with e.g. workers.